Sources/GRPC/Interceptor/ServerInterceptorPipeline.swift

```diff
@@ -128,8 +122,57 @@ internal final class ServerInterceptorPipeline<Request, Response> {
   /// - Important: This *must* to be called from the `eventLoop`.
   @inlinable
   internal func receive(_ part: GRPCServerRequestPart<Request>) {
+    self._invokeReceive(part, onContextAtIndex: self._headIndex)
+  }
+
+  /// Invoke receive on the appropriate context when called from the context at the given index.
+  @inlinable
+  internal func invokeReceive(
+    _ part: GRPCServerRequestPart<Request>,
+    fromContextAtIndex index: Int
+  ) {
+    self._invokeReceive(part, onContextAtIndex: self._nextInboundIndex(after: index))
+  }
+
+  /// Invoke receive on the context at the given index, if doing so is safe.
+  @inlinable
+  internal func _invokeReceive(
+    _ part: GRPCServerRequestPart<Request>,
+    onContextAtIndex index: Int
+  ) {
     self.eventLoop.assertInEventLoop()
-    self.head?.invokeReceive(part)
+    assert(index >= self._headIndex)
+    assert(index <= self._tailIndex)
+
+    guard self._isOpen else {
+      return
+    }
+
+    // We've checked the index.
+    self._invokeReceive(part, onContextAtUncheckedIndex: index)
+  }
+
+  /// Invoke receive on the context at the given index, assuming that the index is valid and the
+  /// pipeline is still open.
+  @inlinable
+  internal func _invokeReceive(
+    _ part: GRPCServerRequestPart<Request>,
+    onContextAtUncheckedIndex index: Int
+  ) {
+    switch index {
+    case self._headIndex:
+      // The next inbound index must exist, either for the tail or a user interceptor.
+      self._invokeReceive(
+        part,
+        onContextAtUncheckedIndex: self._nextInboundIndex(after: self._headIndex)
+      )
+
+    case self._tailIndex:
+      self._onRequestPart(part)
+
+    default:
+      self._userContexts[index].interceptor.receive(part, context: self._userContexts[index])
```

Comment:
This seems like a slightly tricky spelling: I don't love that we have to subscript the array twice. Can this be pushed into a method on the context?